include/unifex/repeat.hpp

```diff
@@ -0,0 +1,235 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/type_traits.hpp>
+#include <unifex/type_list.hpp>
+#include <unifex/manual_lifetime.hpp>
+#include <unifex/manual_lifetime_union.hpp>
+#include <unifex/async_trace.hpp>
+
+#include <utility>
+#include <cassert>
+#include <exception>
+
+namespace unifex {
+
+namespace detail {
+
+template<typename Source, typename Receiver>
+class repeat_operation;
+
+template<typename Source, typename Receiver>
+class repeat_source_receiver;
+
+template<typename Source, typename Receiver>
+class repeat_source_receiver {
+  using operation = repeat_operation<Source, Receiver>;
+public:
+  explicit repeat_source_receiver(operation* op) noexcept
+  : op_(op) {}
+
+  repeat_source_receiver(repeat_source_receiver&& other) noexcept
+  : op_(std::exchange(other.op_, {}))
+  {}
+ 
+  void set_value() noexcept {
+    assert(op_ != nullptr);
+
+    // This signals to repeat the operation.
+    auto* op = op_;
+
+    if (op->isSourceOpConstructed_) {
+      op->sourceOp_.destruct();
+      op->isSourceOpConstructed_ = false;
+    }
+
+    if constexpr (is_nothrow_connectable_v<Source&, repeat_source_receiver>) {
+      auto& sourceOp = op->sourceOp_.construct_from([&]() noexcept {
+          return unifex::connect(op->source_, repeat_source_receiver{op});
+        });
+      op->isSourceOpConstructed_ = true;
+      unifex::start(sourceOp);
+    } else {
+      try {
+        auto& sourceOp = op->sourceOp_.construct_from([&] {
+            return unifex::connect(op->source_, repeat_source_receiver{op});
+          });
+        op->isSourceOpConstructed_ = true;
+        unifex::start(sourceOp);
+      } catch (...) {
+        unifex::set_error((Receiver&&)op->receiver_, std::current_exception());
+      }
+    }
+  }
+
+  template<
+    typename R = Receiver,
+    std::enable_if_t<std::is_invocable_v<decltype(unifex::set_done), R>, int> = 0>
+  void set_done() noexcept {
+    assert(op_ != nullptr);
+    unifex::set_done(std::move(op_->receiver_));
+  }
+
+  template<
+    typename Error,
+    std::enable_if_t<std::is_invocable_v<decltype(unifex::set_error), Receiver, Error>, int> = 0>
+  void set_error(Error&& error) noexcept {
+    assert(op_ != nullptr);
+    unifex::set_error(std::move(op_->receiver_), (Error&&)error);
+  }
+
+private:
+  template<typename CPO, typename... Args>
+  friend auto tag_invoke(CPO cpo, const repeat_source_receiver& r, Args&&... args)
+      noexcept(std::is_nothrow_invocable_v<CPO, const Receiver&, Args...>)
+      -> std::invoke_result_t<CPO, const Receiver&, Args...> {
+    return std::move(cpo)(r.get_receiver(), (Args&&)args...);
+  }
```

Comment:
This will probably require an `enable_if<is_callable_v<CPO, const Receiver&, Args...>, int> = 0` to be added here to avoid potential for recursive instantiation under msvc.